### PR TITLE
Owls 87209 - Domain lifecycle scripts fix in cases where domain/cluster specs are overridden at the individual managed server level

### PR DESCRIPTION
**Pull request by ankedia**
Owls 87209 - Domain lifecycle scripts fix in cases where domain/cluster specs are overridden at the individual managed server level. The script doesn't work correctly when `.spec.managedServers` section has pre-existing details for a particular managed server. For e.g Consider the below stanza where the `serverStartState` attribute is overridden for soa_server1.
spec:
  managedServers:    
  - serverName: soa_server1
    serverStartState: RUNNING

This change adds the server start policy element for the managed server if the policy doesn't exist or replaces the policy if the policy already exists. Similarly, it unsets the policy and/or deletes the managed server element if managed server spec only has serverStartPolicy under it.

### CODE REVIEW

File: kubernetes/samples/scripts/domain-lifecycle/helper.sh [hunk 1]
```diff
@@ -111,14 +111,20 @@ function createServerStartPolicyPatch {
...
-  if [ -z "${currentStartPolicy}" ]; then
+  managedServers=$(echo ${domainJson} | jq -cr '(.spec.managedServers)')
+  if [[ -z "${currentStartPolicy}" && "${managedServers}" == "null" ]]; then
     # Server start policy doesn't exist, add a new policy
     addPolicyCmd=".[.| length] |= . + {\"serverName\":\"${serverName}\", \
       \"serverStartPolicy\":\"${policy}\"}"
     serverStartPolicyPatch=$(echo ${domainJson} | jq .spec.managedServers | jq -c "${addPolicyCmd}")
+  elif [ "${managedServers}" != "null" ]; then
+    extractSpecCmd="(.spec.managedServers)"
+    mapCmd=". |= (map(.serverName) | index (\"${serverName}\")) as \$idx | if \$idx then .[\$idx][\"serverStartPolicy\"] = \"${policy}\" else .+  [{serverName: \"${serverName}\" , serverStartPolicy: \"${policy}\"}] end"
```
Comment (tbarnes-us) [inline kubernetes/samples/scripts/domain-lifecycle/helper.sh:126]: This is a long line. Something like the following would be easier to read:
```
mapCmd="\
  . |= (map(.serverName) | index (\"${serverName}\")) as \$idx | \
  if \$idx \
  then .[\$idx][\"serverStartPolicy\"] = \"${policy}\" \
  else .+  [{serverName: \"${serverName}\" , serverStartPolicy: \"${policy}\"}] end "
```
(I may have a syntax error in above - it's just intended as an example.)

File: kubernetes/samples/scripts/domain-lifecycle/helper.sh [hunk 2]
```diff
@@ -199,13 +203,34 @@ function createPatchJsonToUnsetPolicy {
...
+
+  unsetCmd="(.spec.managedServers[] | select (.serverName == \"${serverName}\") | del (.serverStartPolicy))"
+  replacePolicyCmd=$(echo ${domainJson} | jq -cr "${unsetCmd}")
+  replacePolicyCmdLen=$(echo "${replacePolicyCmd}" | jq -e keys_unsorted | jq length)
+  if [ ${replacePolicyCmdLen} == 1 ]; then
+    mapCmd=". |= map(if .serverName == \"${serverName}\" then del(.) else . end)"
+  else
+    mapCmd=". |= map(if .serverName == \"${serverName}\" then . = ${replacePolicyCmd} else . end)"
+  fi
+  serverStartPolicyPatch=$(echo ${domainJson} | jq "(.spec.managedServers)" | jq "${mapCmd}")
```
Comment (tbarnes-us) [inline kubernetes/samples/scripts/domain-lifecycle/helper.sh:230]: It looks like the intent is for serverStartPolicyPatch to be a local variable, but it has the same name as the variable used for __result by the caller(s).  Maybe give it a different name and declare it with 'local'.

**Comment by tbarnes-us** [comment]
General comment: IIRC, there are integration tests for these scripts. If so, please file a JIRA to enhance them to cover the new use case...

**Review by anpanigr — COMMENTED** [review]
Can you enhance integration test (ItServerStartPolicy.java) to add this usecase.

**Comment by ankedia** [comment]
> General comment: IIRC, there are integration tests for these scripts. If so, please file a JIRA to enhance them to cover the new use case...

Thanks. I have made the suggested changes and also created JIRA OWLS-87819 to track the enhancement of the integration tests to cover the new use case.

**Comment by ankedia** [comment]
> Can you enhance integration test (ItServerStartPolicy.java) to add this usecase.
I have created JIRA OWLS-87819 to track the integration test enhancement and it can be done in a separate PR. Do we have any other integration tests where domain/cluster specs are overridden at the managed server level?

**Comment by tbarnes-us** [comment]
> >Can you enhance integration test (ItServerStartPolicy.java) to add this usecase.
> I have created JIRA OWLS-87819 to track the integration test enhancement and it can be done in a separate PR. 
Thanks.
> Do we have any other integration tests where domain/cluster specs are overridden at the managed server level?
I don't know, Pani might.

**Review by tbarnes-us — APPROVED** [review]

**Review by rjeberhard — APPROVED** [review]

**Pull request merged**